### PR TITLE
Build for windows with CMake

### PR DESCRIPTION
**Pull request by h-vetinari**
Sorry to see the issues you've recently had with CMake. The good news is that this gave me a great starting off point (i.e. I just used the CMake options from #10), because I believe windows should be built with CMake too.

I'm running into issues in https://github.com/conda-forge/sdl2_image-feedstock/pull/12 with the automatic detection of `webp`, because - AFAICT - this feedstock uses non-standard names for the libraries (i.e. normally on windows there's no "lib" prefix).

While I'm add it simplify the test section a bit to make it a bit more consistent and (hopefully) easier to read.

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by h-vetinari** [comment]
It's very interesting IMO that this failed where #13 succeeded on master a mere 3h ago - and I didn't change anything about the unix builds!

_Narrator: he did._

**Comment by carterbox** [comment]
Hi @h-vetinari,

I tried using CMake locally while I was refactoring for the previous build; it wasn't configured correctly to successfully create shared libraries on Windows?

Furthermore...

As I alluded to in [this comment](https://github.com/conda-forge/libwebp-base-feedstock/issues/11#issuecomment-1189287168), the CMake build system doesn't seem to be officially recommended/supported by upstream.

This feedstock currently uses nmake for Windows according to the ["canonical" build instructions](https://developers.google.com/speed/webp/docs/compiling). So if the library name is non-standard, it is the fault of upstream and should not be fixed here (especially since there are so many packages that rely on this library already).

Like with the recent ABI breakage for osx, changing the *name* of the shared library on Windows would be ABI breaking, so we would reject this change.

This PR would only be accepted if the switch to CMake did not change the ABI/names and instead only added CMake build artifacts.

**Comment by h-vetinari** [comment]
hi @carterbox, thanks for the quick response!

> As I alluded to in [this comment](https://github.com/conda-forge/libwebp-base-feedstock/issues/11#issuecomment-1189287168), the CMake build system doesn't seem to be officially recommended/supported by upstream.

I do not think "we tend to treat autoconf/configure as the canonical build system" as a strong enough statement for avoiding cmake, especially given the benefits it has (~and just copying DLLs from the source is really not good practice~ edit: sorry, misread `bld.bat`).

> This feedstock currently uses nmake for Windows according to the ["canonical" build instructions](https://developers.google.com/speed/webp/docs/compiling). So if the library name is non-standard, it is the fault of upstream and should not be fixed here (especially since there are so many packages that rely on this library already).

We could use nmake as the cmake generator, but I care less about upstream defaults than conda-forge internal consistency. And the gateway through the feedstock is a really clean point to make such choices.
 
> Like with the recent ABI breakage for osx, changing the _name_ of the shared library on Windows would be ABI breaking, so we would reject this change.

That would be easy to fix by having symlinks and/or hard-copies for a while (+ perhaps some repodata-patches for builds against old webp)
 
> This PR would only be accepted if the switch to CMake did not change the ABI/names and instead only added CMake build artifacts.

Could you avoid making such definite statements right off the bat, please?

**Comment by carterbox** [comment]
> (~~and just copying DLLs from the source is really not good practice~~ edit: sorry, misread bld.bat).

The scripts use copy because the upstream build system does not support splitting the package between binaries and libraries/headers. I am open to alternative approaches for this problem.

> That would be easy to fix by having symlinks and/or hard-copies for a while (+ perhaps some repodata-patches for builds against old webp)

symlinks require admin permissions on Windows which I don't think conda-build has, you can try though.

> I do not think "we tend to treat autoconf/configure as the canonical build system" as a strong enough statement for avoiding cmake

I agree. That's not the reason. I tried to transition this feedstock to CMake. I think having CMake metadata would be very helpful. The reason is that upstream doesn't ensure that the CMake build system generates the same outputs as the non-CMake build system.

> I care less about upstream defaults than conda-forge internal consistency.

The names the canonical outputs are important in the context of this very popular library because many projects will have configured their build systems to look for the canonically named outputs. And if we care about conda-forge consistency, then we should not change the names of artifacts in already released packages?

> This PR would only be accepted if the switch to CMake did not change the ABI/names and instead only added CMake build artifacts.

Your goal is to find webp with CMake at build time. The constraint is that we don't break the exisiting ABI at runtime. I think this is a reasonable constraint. We can have both 😄; patch the CMakeLists instead of trying to make symlinks or hardcopies.

**Comment by h-vetinari** [comment]
> Your goal is to find webp with CMake at build time. The constraint is that we don't break the exisiting ABI at runtime. I think this is a reasonable constraint. We can have both 😄; patch the CMakeLists instead of trying to make symlinks or hardcopies.

If/once I manage to build with CMake at all, I don't mind patching in a name change on windows.

PS. the upstream [`CMakeLists.txt`](https://chromium.googlesource.com/webm/libwebp/+blame/refs/tags/v1.2.3/CMakeLists.txt) seems quite actively maintained, but I don't know how well they test it on windows. It's a real shame they don't make CMake the reference, instead of maintaining two separate master files already (unix & windows).

**Comment by h-vetinari** [comment]
PTAL @carterbox

I've also sent the patch to the upstream maintainers in the hope that we don't have to carry it forever.

File: recipe/patches/0001-match-naming-convention-used-by-nmake-also-for-windo.patch
Comment (carterbox) [inline recipe/patches/0001-match-naming-convention-used-by-nmake-also-for-windo.patch]:
```cmake
if(WIN32)
    # Prefix all shared libraries with 'lib'.
    set(CMAKE_SHARED_LIBRARY_PREFIX "lib")
    # Prefix all static libraries with 'lib'.
    set(CMAKE_STATIC_LIBRARY_PREFIX "lib")
endif()
```
Were you looking for something like this? You want all of the libraries to be prefixed with "lib", right?

source: https://stackoverflow.com/a/63471388

**Review by carterbox — CHANGES_REQUESTED** [review]
Just one suggestion for now. I'll check the log and outputs later.

File: recipe/patches/0001-match-naming-convention-used-by-nmake-also-for-windo.patch
Comment (h-vetinari) [reply recipe/patches/0001-match-naming-convention-used-by-nmake-also-for-windo.patch]: Yeah, that's nicer, I agree. Googling the right CMake variable is hard.

Comment (carterbox) [reply recipe/patches/0001-match-naming-convention-used-by-nmake-also-for-windo.patch]: > Googling the right CMake variable is hard.

Agreed.

Comment (h-vetinari) [reply recipe/patches/0001-match-naming-convention-used-by-nmake-also-for-windo.patch]: So, are we good to go here now? 🙃

**Review by carterbox — CHANGES_REQUESTED** [review]
CMake files are usually installed to `Library\lib\cmake\WebP`? Not `Library\share\WebP\cmake`? (To be clear, it doesn't really matter to me as long as CMAKE find_package works!)

Other than that, I think the libraries are compatible. The addition of EXPORT_ALL_SYMBOLS=ON causes more symbols to be exported than in the old library, but all the old symbols seem to be there which is what's important?

I also tested the current webp binaries with this new library build and they worked.

**Comment by h-vetinari** [comment]
> CMake files are usually installed to `Library\lib\cmake\WebP`? Not `Library\share\WebP\cmake`? (To be clear, it doesn't really matter to me as long as CMAKE find_package works!)

AFAIK both /lib and /share are in the CMake module search path. I'd let CMake figure that one out.



> Other than that, I think the libraries are compatible. The addition of EXPORT_ALL_SYMBOLS=ON causes more symbols to be exported than in the old library, but all the old symbols seem to be there which is what's important?

There's no other simple way to get those symbols into a .lib without adding `__declspec(dllexport)` _everywhere_, so I'd say this is a fair trade-off.

> I also tested the current webp binaries with this new library build and they worked.

Great, thanks! Sounds like this should be finished then, or what changes do you still envision?

**Review by carterbox — APPROVED** [review]
I tested `find_package(WebP 1 REQUIRED)` and it seems to locate webp, so LGTM.

**Comment by carterbox** [comment]
@conda-forge-admin, please rerender

**Comment by h-vetinari** [comment]
Thanks for the review & merge!

**Comment by h-vetinari** [comment]
> I tested `find_package(WebP 1 REQUIRED)` and it seems to locate webp, so LGTM.

Weeeeeeeeeeeell shucks. `find_package(webp)` doesn't works on windows at all, and switching to `find_package(WebP)` still [fails](https://github.com/conda-forge/sdl2_image-feedstock/pull/12) shortly after:
```
CMake Warning (dev) at D:/bld/sdl2_image_1658511832678/_build_env/Library/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.cmake:438 (message):
  The package name passed to `find_package_handle_standard_args` (webp) does
  not match the name of the calling package (WebP).  This can lead to
  problems in calling code that expects `find_package` result variables
  (e.g., `_FOUND`) to follow a certain pattern.
```
(and is incompatible with linux apparently, but not OSX for some reason).

The windows error says further:
```
  Could not find a package configuration file provided by "WebP" with any of
  the following names:

    WebPConfig.cmake
    webp-config.cmake
```

but what gets installed as per this PR is:
```
-- Installing: %PREFIX%\Library/share/WebP/cmake/WebPTargets.cmake
-- Installing: %PREFIX%\Library/share/WebP/cmake/WebPTargets-release.cmake
-- Installing: %PREFIX%\Library/share/WebP/cmake/WebPConfigVersion.cmake
-- Installing: %PREFIX%\Library/share/WebP/cmake/WebPConfig.cmake
```
Upon further inspection, this is probably influenced by the following direct [settings](https://chromium.googlesource.com/webm/libwebp/+blame/refs/tags/v1.2.3/CMakeLists.txt#732-L750) in the upstream `CMakeLists.txt`

**Comment by h-vetinari** [comment]
It seems we're _just_ not falling into the search [behaviour](https://cmake.org/cmake/help/latest/command/find_package.html#config-mode-search-procedure) of `find_package`:
> The tables below show the directories searched. Each entry is meant for installation trees following Windows (W), UNIX (U), or Apple (A) conventions:
```
<prefix>/                                                       (W)
<prefix>/(cmake|CMake)/                                         (W)
<prefix>/<name>*/                                               (W)
<prefix>/<name>*/(cmake|CMake)/                                 (W)
<prefix>/(lib/<arch>|lib*|share)/cmake/<name>*/                 (U)
<prefix>/(lib/<arch>|lib*|share)/<name>*/                       (U)
<prefix>/(lib/<arch>|lib*|share)/<name>*/(cmake|CMake)/         (U)
<prefix>/<name>*/(lib/<arch>|lib*|share)/cmake/<name>*/         (W/U)
<prefix>/<name>*/(lib/<arch>|lib*|share)/<name>*/               (W/U)
<prefix>/<name>*/(lib/<arch>|lib*|share)/<name>*/(cmake|CMake)/ (W/U)
```

**Comment by carterbox** [comment]
This is what I did to test it.

```
conda create -n webp libwebp cmake ninja vs2017_win-64
```

Then use cmake to configure this project

```
project(testbuild)

find_package(WebP 1 REQUIRED )
```

**Pull request merged**
